File: nodejs/lib/bit-dict.js
'use strict';

/**
 * BitDict - N비트 전수 사전 관리
 *
 * MSBD 파일 포맷 (12B 헤더):
 *   magic(4B 'MSBD') + version(2B) + bitDepth(2B) + reserved(4B)
 *
 * index = pattern value (수학적 항등) → 실제 엔트리 데이터 불필요
 */

const fs = require('fs');
const os = require('os');
const path = require('path');

const MSBD_MAGIC = Buffer.from('MSBD');
const MSBD_VERSION = 1;
const MSBD_HEADER_SIZE = 12;
const DEFAULT_DICT_DIR = path.join(os.homedir(), '.mshzip');

class BitDict {
  /**
   * @param {Object} [opts]
   * @param {string} [opts.dictDir] - 사전 디렉토리
   * @param {number} [opts.maxMemBytes] - 최대 메모리 (기본: freemem × 0.8)
   */
  constructor(opts = {}) {
    this.dir = opts.dictDir || process.env.MSHZIP_DICT_DIR || DEFAULT_DICT_DIR;
    this.maxMemBytes = opts.maxMemBytes || Math.floor(os.freemem() * 0.8);
  }

  /**
   * MSBD 파일 경로
   * @param {number} bitDepth
   * @returns {string}
   */
  path(bitDepth) {
    return path.join(this.dir, `bitdict-${bitDepth}.msbd`);
  }

  /**
   * MSBD 사전 파일 생성
   * @param {number} bitDepth - 비트 깊이 (1~32)
   * @returns {string} 생성된 파일 경로
   */
  generate(bitDepth) {
    if (bitDepth < 1 || bitDepth > 32) {
      throw new Error(`bitDepth out of range: 1~32, got ${bitDepth}`);
    }
    this._ensureDir();
    const filePath = this.path(bitDepth);
    const header = Buffer.alloc(MSBD_HEADER_SIZE, 0);
    MSBD_MAGIC.copy(header, 0);
    header.writeUInt16LE(MSBD_VERSION, 4);
    header.writeUInt16LE(bitDepth, 6);
    // reserved(4B) = 0 (이미 alloc으로 초기화)
    fs.writeFileSync(filePath, header);
    return filePath;
  }

  /**
   * 사전 로드 (없으면 자동 생성)
   * @param {number} bitDepth
   * @returns {number} 로드된 bitDepth
   */
  load(bitDepth) {
    const filePath = this.path(bitDepth);
    if (!fs.existsSync(filePath)) {
      this.generate(bitDepth);
      return bitDepth;
    }

    const buf = fs.readFileSync(filePath);
    if (buf.length < MSBD_HEADER_SIZE) {
      throw new Error(`MSBD 파일 손상: ${filePath}`);
    }
    if (!buf.slice(0, 4).equals(MSBD_MAGIC)) {
      throw new Error('MSBD 매직넘버 불일치');
    }
    const fileBitDepth = buf.readUInt16LE(6);
    if (fileBitDepth !== bitDepth) {
      throw new Error(`bitDepth 불일치: 파일=${fileBitDepth}, 요청=${bitDepth}`);
    }
    return bitDepth;
  }

  /**
   * N비트 전수 사전의 추정 메모리 (바이트)
   * @param {number} bitDepth
   * @returns {number}
   */
  estimateMemBytes(bitDepth) {
    return Math.pow(2, bitDepth) * 4;
  }

  /**
   * 시스템 메모리 초과 여부
   * @param {number} bitDepth
   * @returns {boolean}
   */
  isOverLimit(bitDepth) {
    return this.estimateMemBytes(bitDepth) >= this.maxMemBytes;
  }

  /**
   * 사전 정보 조회
   * @param {number} bitDepth
   * @returns {Object}
   */
  info(bitDepth) {
    const filePath = this.path(bitDepth);
    const result = {
      exists: false,
      path: filePath,
      bitDepth,
      patternCount: Math.pow(2, bitDepth),
      estimatedMemMB: +(this.estimateMemBytes(bitDepth) / 1024 / 1024).toFixed(1),
      isOverLimit: this.isOverLimit(bitDepth),
    };

    if (!fs.existsSync(filePath)) return result;

    try {
      const buf = Buffer.alloc(MSBD_HEADER_SIZE);
      const fd = fs.openSync(filePath, 'r');
      fs.readSync(fd, buf, 0, MSBD_HEADER_SIZE, 0);
      fs.closeSync(fd);
      result.exists = true;
    } catch { /* 읽기 실패 */ }

    return result;
  }

  /** @private */
  _ensureDir() {
    if (!fs.existsSync(this.dir)) {
      fs.mkdirSync(this.dir, { recursive: true });
    }
  }
}

module.exports = { BitDict, MSBD_MAGIC, MSBD_VERSION, MSBD_HEADER_SIZE };

File: python/src/mshzip/bit_dict.py
'BitDict - N비트 전수 사전 관리'
from __future__ import annotations

import os
import struct
from pathlib import Path

from .dict_store import _get_free_mem

# MSBD 파일 포맷 (12B 헤더):
#   magic(4B 'MSBD') + version(2B) + bitDepth(2B) + reserved(4B)
MSBD_MAGIC: bytes = b'MSBD'
MSBD_VERSION: int = 1
MSBD_HEADER_SIZE: int = 12

DEFAULT_DICT_DIR: str = str(Path.home() / '.mshzip')


class BitDict:
    """N비트 전수 사전 관리.

    index = pattern value (수학적 항등) -> 실제 엔트리 데이터 불필요.
    MSBD 파일은 bitDepth 메타데이터만 저장.
    """

    def __init__(
        self,
        dict_dir: str | None = None,
        max_mem_bytes: int | None = None,
    ) -> None:
        self.dir = dict_dir or os.environ.get('MSHZIP_DICT_DIR', DEFAULT_DICT_DIR)
        self.max_mem_bytes = max_mem_bytes or int(_get_free_mem() * 0.8)

    def path(self, bit_depth: int) -> str:
        """MSBD 파일 경로."""
        return os.path.join(self.dir, f'bitdict-{bit_depth}.msbd')

    def generate(self, bit_depth: int) -> str:
        """MSBD 사전 파일 생성."""
        if bit_depth < 1 or bit_depth > 32:
            raise ValueError(f'bit_depth out of range: 1~32, got {bit_depth}')

        self._ensure_dir()
        file_path = self.path(bit_depth)

        header = bytearray(MSBD_HEADER_SIZE)
        header[0:4] = MSBD_MAGIC
        struct.pack_into('<H', header, 4, MSBD_VERSION)
        struct.pack_into('<H', header, 6, bit_depth)
        # reserved(4B) = 0 (이미 bytearray로 초기화)

        with open(file_path, 'wb') as f:
            f.write(header)

        return file_path

    def load(self, bit_depth: int) -> int:
        """사전 로드 (없으면 자동 생성)."""
        file_path = self.path(bit_depth)
        if not os.path.exists(file_path):
            self.generate(bit_depth)
            return bit_depth

        with open(file_path, 'rb') as f:
            buf = f.read(MSBD_HEADER_SIZE)

        if len(buf) < MSBD_HEADER_SIZE:
            raise ValueError(f'MSBD 파일 손상: {file_path}')

        if buf[0:4] != MSBD_MAGIC:
            raise ValueError(f'MSBD 매직넘버 불일치: {buf[0:4]!r}')

        file_bit_depth = struct.unpack_from('<H', buf, 6)[0]
        if file_bit_depth != bit_depth:
            raise ValueError(
                f'bitDepth 불일치: 파일={file_bit_depth}, 요청={bit_depth}'
            )

        return bit_depth

    def estimate_mem_bytes(self, bit_depth: int) -> int:
        """N비트 전수 사전의 추정 메모리 (바이트)."""
        return (2 ** bit_depth) * 4

    def is_over_limit(self, bit_depth: int) -> bool:
        """시스템 메모리 초과 여부."""
        return self.estimate_mem_bytes(bit_depth) >= self.max_mem_bytes

    def info(self, bit_depth: int) -> dict:
        """사전 정보 조회."""
        file_path = self.path(bit_depth)
        result: dict = {
            'exists': False,
            'path': file_path,
            'bit_depth': bit_depth,
            'pattern_count': 2 ** bit_depth,
            'estimated_mem_mb': round(
                self.estimate_mem_bytes(bit_depth) / 1024 / 1024, 1
            ),
            'is_over_limit': self.is_over_limit(bit_depth),
        }

        if not os.path.exists(file_path):
            return result

        try:
            with open(file_path, 'rb') as f:
                f.read(MSBD_HEADER_SIZE)
            result['exists'] = True
        except OSError:
            pass

        return result

    def _ensure_dir(self) -> None:
        os.makedirs(self.dir, exist_ok=True)
